### DISCUSSION TITLE
How do I properly type a `GuardMeta`?

### DISCUSSION THREAD

**Discussion by chopfitzroy** [statelyai/xstate#3373] [Q&A]
Largely inspired by #414 I am trying to create an `every` function to check against multiple conditions in a guard.

Now I have this _working_ in terms of functionality but I am having a really hard time satisfying the type checker.

Here is a working [sandbox](https://codesandbox.io/s/old-pine-tin8pr?file=/src/index.ts ) of what I have (with type errors).

The type errors I am talking about are in `machineUtils.ts`. The type I am struggling with is the `EveryGuardMeta` which is currently defined twice in the file (one commented out) as both:

```ts
type EveryGuardMeta<Context> = GuardMeta<Context, AnyEventObject>;
```

And the version I would like to get working:
```ts
interface EveryGuardEvent extends EventObject {
    guardsToCheckAgainst: string[];
}

type EveryGuardMeta<Context> = GuardMeta<Context, EveryGuardEvent>;
```

What I would like is for this function:
```ts
const hasEveryGuard =
    <Context>(existingGuards: Guards<Context>) =>
    (context: Context, event: AnyEventObject, meta: EveryGuardMeta<Context>): boolean => {
        const { guardsToCheckAgainst = [] } = meta.cond;
        return guardsToCheckAgainst.every((key) => existingGuards[key](context, event, meta));
    };
```

To correctly understand the `meta.cond` and know that `guardsToCheckAgainst` is a `string[]`.

Any help with this would be greatly appreciated I have searched high and low for the correct types to use here.

**Comment by Andarist** [comment]
This is rather a complex situation and ultimately I'm not sure if you can achieve what you want in a truly e2e type-safe manner. I would just recommend using some extra casts in v4. The good news though is that guards like `every`, `not` and similar will be provided by the upcoming v5 so you won't have to reinvent the wheel on that front.

**Reply by chopfitzroy** [reply]
Fantastic! And thank you for your response!
